File: OpenMediationDemo/WelcomeViewController.m
// Copyright 2020 ADTIMING TECHNOLOGY COMPANY LIMITED
// Licensed under the GNU Lesser General Public License Version 3

#import "WelcomeViewController.h"
#import "MainViewController.h"
#import "InitSettingViewController.h"
#import <AppTrackingTransparency/AppTrackingTransparency.h>
#import <AdSupport/AdSupport.h>

@import OpenMediation;

@interface WelcomeViewController ()
@property (strong, nonatomic) UILabel *welcomLabel;
@property (strong, nonatomic) UIButton *settingBtn;
@end

@implementation WelcomeViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    
    UIBarButtonItem *showBar = [[UIBarButtonItem alloc]initWithTitle:@"Ad" style:UIBarButtonItemStylePlain target:self action:@selector(showItemAction)];
    self.navigationItem.rightBarButtonItem = showBar;
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(initSuccess) name:@"kOpenMediatonInitSuccessNotification" object: nil];
    [self.view addSubview:self.welcomLabel];
    [self.view addSubview:self.settingBtn];
    
    UIButton *idfaBtn = [UIButton buttonWithType:UIButtonTypeSystem];
    idfaBtn = [[UIButton alloc] initWithFrame:CGRectMake(0, self.view.frame.size.height/3.0*2.0 - 100,self.view.frame.size.width, 50)];
    idfaBtn.titleLabel.font = [UIFont systemFontOfSize:15];
    [idfaBtn setTitleColor:[UIColor systemBlueColor] forState:UIControlStateNormal];
    [idfaBtn setTitle:@"RequestIDFAAuthorization" forState:UIControlStateNormal];
    [idfaBtn addTarget:self action:@selector(requestIDFAAuthorization) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:idfaBtn];
}

- (void)requestIDFAAuthorization {
    if (@available(iOS 14, *)) {
        [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
            if (status == ATTrackingManagerAuthorizationStatusAuthorized) {
                NSString *idfa = [[ASIdentifierManager sharedManager].advertisingIdentifier UUIDString];
                NSLog(@"IDFA: %@",idfa);
            } else {
                NSLog(@"The authorization status %ud",(unsigned int)status);
            }
        }];
    }
}

- (void)initSuccess {
    self.welcomLabel.text = [self.welcomLabel.text stringByAppendingString:@"\n\nOpenMediation init success!!!"];
}

- (UILabel*)welcomLabel {
    if (!_welcomLabel) {
        _welcomLabel = [[UILabel alloc]initWithFrame:CGRectMake(0 ,150,self.view.frame.size.width,200)];
        _welcomLabel.textAlignment = NSTextAlignmentCenter;
        _welcomLabel.numberOfLines = 0;
        NSString *buildDate = [NSString stringWithFormat:@"%s %s",__DATE__, __TIME__];
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        dateFormatter.locale = [[NSLocale alloc] initWithLocaleIdentifier:@"en_US"];
        dateFormatter.dateFormat = @"MMM dd yyyy HH:mm:ss";
        NSDate *date = [dateFormatter dateFromString:buildDate];
        [dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
        dateFormatter.timeZone = [NSTimeZone systemTimeZone];
        NSString* dateStr = [dateFormatter stringFromDate:date];
        _welcomLabel.text = [NSString stringWithFormat:@"V%@\n\nBuild: %@",[OpenMediation SDKVersion],dateStr];
    }
    return _welcomLabel;
}

- (void)showItemAction {
    MainViewController *listVC = [[MainViewController alloc] init];
    [self.navigationController pushViewController:listVC animated:YES];
}

- (UIButton*)settingBtn {
    if(!_settingBtn){
        _settingBtn = [[UIButton alloc] initWithFrame:CGRectMake(0, self.view.frame.size.height/3.0*2.0,self.view.frame.size.width, 50)];
        _settingBtn.titleLabel.font = [UIFont systemFontOfSize:15];
        [_settingBtn setTitleColor:[UIColor systemBlueColor] forState:UIControlStateNormal];
        [_settingBtn setTitle:@"Init Settings" forState:UIControlStateNormal];
        [_settingBtn addTarget:self action:@selector(setting) forControlEvents:UIControlEventTouchUpInside];
    }
    return _settingBtn;
}

- (void)setting {
    InitSettingViewController *settingVC = [[InitSettingViewController alloc] init];
    [self.navigationController pushViewController:settingVC animated:YES];
}
@end

File: OpenMediationSwiftDemo/WelcomeViewController.swift
// Copyright 2020 ADTIMING TECHNOLOGY COMPANY LIMITED
// Licensed under the GNU Lesser General Public License Version 3

import UIKit
import AppTrackingTransparency
import OpenMediation
import FBAudienceNetwork
import AdSupport

var welcomeLab = UILabel()

class WelcomeViewController: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let itemBtn = UIBarButtonItem.init(title: "Ad", style: .plain, target: self, action: #selector(showItemAction))
        self.navigationItem.rightBarButtonItem = itemBtn
        
        var idfaBtn = UIButton.init(type: UIButton.ButtonType.system)
        idfaBtn = UIButton(frame: CGRect(x: 0, y: self.view.frame.size.height/3.0*2.0 - 100, width: self.view.frame.size.width, height: 50))
        idfaBtn.setTitleColor(UIColor.systemBlue, for: .normal)
        idfaBtn.titleLabel?.font = UIFont.systemFont(ofSize: 15)
        idfaBtn.setTitle("RequestIDFAAuthorization", for: .normal)
        idfaBtn.addTarget(self, action: #selector(idfaBtnDidClicked), for: .touchUpInside)
        self.view.addSubview(idfaBtn)
        
        NotificationCenter.default.addObserver(self, selector: #selector(initSuccess), name: Notification.Name("kOpenMediatonInitSuccessNotification"), object: nil)
        
        
        welcomeLab = UILabel.init(frame: CGRect(x: 0, y: 200, width: self.view.frame.size.width, height: 200))
        welcomeLab.textAlignment = .center
        welcomeLab.numberOfLines = 0
        
        let buildDate = NSString()
        let dateformatter = DateFormatter()
        dateformatter.locale = NSLocale.init(localeIdentifier: "en_US") as Locale
        dateformatter.dateFormat = "MMM dd yyyy HH:mm:ss"
        let date = dateformatter.date(from: buildDate as String) ?? Date.init()
        dateformatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        dateformatter.timeZone = NSTimeZone.system
        let dateStr = dateformatter.string(from: date)
        let version: String = OpenMediation.sdkVersion()
        
        welcomeLab.text = "V" + version + "\n\n" + dateStr
        self.view.addSubview(welcomeLab)
        
    }
    
    @objc func initSuccess() {
        welcomeLab.text = welcomeLab.text?.appending("\n\nOpenMediation init success!!!")
    }
    
    @objc func showItemAction() {
        self.navigationController?.pushViewController(ViewController(), animated: true)
        
        
    }
    
    @objc func idfaBtnDidClicked() {
        if #available(iOS 14, *) {
            ATTrackingManager.requestTrackingAuthorization { (status) in
                let anyClass: AnyClass? = NSClassFromString("FBAdSettings")
                guard let typeClass = anyClass as? FBAdSettings.Type else { return }
                let selector = NSSelectorFromString("setAdvertiserTrackingEnabled:")
                if typeClass.responds(to: selector) == true {
                    typeClass.setAdvertiserTrackingEnabled(status == .authorized)
                    if status == .authorized {
                        let idfa:String = ASIdentifierManager.shared().advertisingIdentifier.uuidString
                        print("IDFA: %@",idfa)
                    }else{
                        print("The authorization status %@",status)
                    }
                }
            }
        }
    }
    
}
